### DISCUSSION TITLE
Kafka compatible client API

### DISCUSSION THREAD

**Discussion by AshleySchaeffer** [apache/iggy#6] [Ideas]
Hello!

Firstly, Iggy looks great. Well played!

I've (sadly) spent quite a bit of time looking at message queues and streaming platforms like this and what I can't get away from is the Kafka client API. Because of it's widespread adoption there are countless projects and tools that are compatible with it. 

I've not done too much of a deep dive into the code base yet, but it would be good to understand what the appetite for a Kafka compatible client API would be? Based purely on the features list, it implies that Iggy uses a similar internal architecture so I question whether it'd require much work beyond the API itself.

Besides tooling, iggy would benefit from battle tested client libraries in numerous languages, while also being entirely familiar with developers. If a Kafka client API worked perfectly, it could even be a drop in replacement.

**Comment by spetz** [comment]
Hey,

Once again, thank you for the kind words and the interesting idea.  While we already had a discussion last week on our [Discord](https://iggy.rs/discord), I'd also like to sum up a few things in this topic.

Kafka is an extremely popular, mature and reliable tool, thus it could be really helpful and useful for all the developers out there already using Kafka, to provide a client compatible API on the Iggy side.

I think that it wouldn't be too difficult (to be verified of course) to implement the most important methods being sending and polling the messages, as some of the concepts, patterns as well as the underlying data storage are quite similar. However, the Kafka API is much bigger than these two, and it's hard to say at this point if it would be possible to provide the seamless integration with all the available features - certainly, we wouldn't like the developers to experience some undefined behavior or weird errors being returned on our side, since we couldn't provide a 100% compatible SDK.

I'd like this to remain an open topic and encourage anyone willing to help with this subject (research, ideas, PoC etc.) - I would be more than happy to support you with this task :)

**Comment by vivekkoya** [comment]
I am willing to work on this API. What versions should we focus on? 

- [Producer API](https://docs.confluent.io/kafka/kafka-apis.html#producer-api)
- [Consumer API](https://docs.confluent.io/kafka/kafka-apis.html#consumer-api)
- [Admin Client API](https://docs.confluent.io/kafka/kafka-apis.html#admin-api)
- [Connect API](https://docs.confluent.io/kafka/kafka-apis.html#connect-api)
- [Kafka Streams API](https://docs.confluent.io/kafka/kafka-apis.html#streams-api)

**Reply by spetz** [reply]
To begin with, it'd be great to try building the basic producer & consumer. Thanks!

**Reply by AshleySchaeffer** [reply]
I'm also down to contribute.

**Reply by spetz** [reply]
Great to hear, feel free to ask about any questions, either here or on our Discord :)

**Reply by vivekkoya** [reply]
I am sketching out what needs to be done in Iggy in order to support Kafka clients (Producer/Consumer). The relevant files are in -     `iggy/core/binary_protocol/src/client/` correct? I am also new to Rust and am figuring how the integration should be implemented with existing code. Please feel free to share feedback. Thanks
https://excalidraw.com/#room=663b8b53e79e2b344f10,OnKqiu6gF0jPbt0FWLgBbA

**Reply by spetz** [reply]
Yes, that's correct. One of the challenges will be the way of dealing with the `stream` - it's our own concept, which essentially groups multiple topics (just like the topic groups N partitions). Stream can be used e.g. for multi-tenancy or for any other purpose.

**Reply by vivekkoya** [reply]
That's nice, multi-tenancy is an advantage Apache Pulsar has over Kafka.

**Reply by hubcio** [reply]
@AshleySchaeffer @vivekkoya how's progress? do you need any help from our side? have you joined our discord?

**Reply by vivekkoya** [reply]
I am working on the client side API (consumer) implementation.

**Reply by sujitn** [reply]
this would be a good addition, anyone still working on this. i am happy to take a stab at it

**Reply by hubcio** [reply]
I don't think anyone is working on it. @sujitn feel free to take a stab at it, preferably once you have some design in mind, send it here, to this gh discussion.

**Comment by sujitn** [comment]
Not sure if discord or this the correct forum for discuss. I have this initial WIP to check on approach

https://github.com/apache/iggy/pull/3038

**Reply by ryerraguntla** [reply]
@sujitn @vivekkoya @AshleySchaeffer - Please look at the  [3252](https://github.com/apache/iggy/discussions/3252) and [3253](https://github.com/apache/iggy/discussions/3253). I am more than happy to collaborate with all of you.

**Comment by ryerraguntla** [comment]
@sujitn @vivekkoya @AshleySchaeffer - I have published tentative implementation plan in the comments section of https://github.com/apache/iggy/issues/3253 and started off with issue - https://github.com/apache/iggy/issues/3421. Please feel free to jump in and pick up the ones of your interest.
